pkg/channel/event_dispatcher.go

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2018 The Knative Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package channel
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	cloudevents "github.com/cloudevents/sdk-go"
+	cehttp "github.com/cloudevents/sdk-go/pkg/cloudevents/transport/http"
+	"go.opencensus.io/plugin/ochttp/propagation/b3"
+	"go.opencensus.io/trace"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"knative.dev/eventing/pkg/kncloudevents"
+	"knative.dev/eventing/pkg/utils"
+)
+
+const correlationIDHeaderName = "Knative-Correlation-Id"
+
+type Dispatcher interface {
+	// DispatchEvent dispatches an event to a destination over HTTP.
+	//
+	// The destination and reply are DNS names. For names with a single label,
+	// the default namespace is used to expand it into a fully qualified name
+	// within the cluster.
+	DispatchEvent(ctx context.Context, event cloudevents.Event, destination, reply string, defaults DispatchDefaults) error
+}
+
+// EventDispatcher is the 'real' Dispatcher used everywhere except unit tests.
+var _ Dispatcher = &EventDispatcher{}
+
+var propagation = &b3.HTTPFormat{}
+
+// EventDispatcher dispatches events to a destination over HTTP.
+type EventDispatcher struct {
+	ceClient         cloudevents.Client
+	supportedSchemes sets.String
+
+	logger *zap.Logger
+}
+
+// DispatchDefaults provides default parameter values used when dispatching an event.
+type DispatchDefaults struct {
```

Comment:
Can we remove this? Does any code actually use it?

Reply:
It is being used to get the namespace in many places... E.g., event_dispatcher.resolveURL.

Reply:
Thanks for the context. Removed!